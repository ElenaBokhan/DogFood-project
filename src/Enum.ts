export enum ETestId {
    ADD_PRODUCT_NAME_INPUT = 'addProductNameInput',
    ADD_PRODUCT_PRICE_INPUT = 'addProductPriceInput',
    ADD_PRODUCT_DESCRIPTION_INPUT = 'addProductDescriptionInput',
    ADD_PRODUCT_DISCOUNT_INPUT = 'addProductDiscountInput',
    ADD_PRODUCT_STOCK_INPUT = 'addProductStockInput',
    ADD_PRODUCT_WIGHT_INPUT = 'addProductWightInput',
    ADD_PRODUCT_PICTURES_INPUT = 'addProductPicturesInput',
    ADD_PRODUCT_SUBMIT_BUTTON = 'addProductSubmitButton',

    ADD_REVIEW_SUBMIT_BUTTON = 'addReviewSubmitButton',
    ADD_REVIEW_TEXT_INPUT = 'addReviewTextInput',

    ADD_TO_FAVOURITES = 'addToFavouritesButton',
    REMOVE_FROM_FAVOURITES = 'removeFromFavourites',

    BREADCRUMBS_BUTTON = 'breadcrumbsButton',

    FOOTER_MAIN_LOGO = 'footerMainLogo',
    FOOTER_MENU = 'footerMenu',
    FOOTER_CONTACTS = 'footerContacts',

    HEADER_MAIN_LOGO = 'headerMainLogo',
    HEADER_ADD_NEW_ICON = 'headerAddNewProduct',
    HEADER_FAVOURITES_ICON = 'headerFavouritesIcon',
    HEADER_CART_ICON = 'headerCartIcon',
    HEADER_PROFILE_ICON = 'headerProfileIcon',
    HEADER_LOG_OUT_ICON = 'headerLogOutIcon',
    HEADER_SEARCH_INPUT = 'headerSearchInput',

    SIGN_FORM_EMAIL_INPUT = 'signFormEmailInput',
    SIGN_FORM_PASSWORD_INPUT = 'signFormPasswordlInput',
    SIGN_FORM_SUBMIT_BUTTON = 'signFormSubmitButton',

    TITLE_PAGE = 'titlePage',
    NOT_FOUND = 'notFound',

    PRODUCT_LIKE_BUTTON = 'likeButton',
    PRODUCT_TRASH_BUTTON = 'trashButton',
    PRODUCT_NAME = 'productName',
    PRODUCT_IMAGE = 'productImage',
}
