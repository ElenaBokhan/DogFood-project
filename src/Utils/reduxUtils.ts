import type {AnyAction, AsyncThunk} from '@reduxjs/toolkit';

type GenericAsyncThunk = AsyncThunk<unknown, unknown, any>;
type PendingAction = ReturnType<GenericAsyncThunk['pending']>;
type RejectedAction = ReturnType<GenericAsyncThunk['rejected']>;
type FulfilledAction = ReturnType<GenericAsyncThunk['fulfilled']>;

const hasPrefix = (action: AnyAction, prefix: string): boolean => action.type.startsWith(prefix);
const isPending = (action: PendingAction): boolean => action.type.endsWith('/pending');
const isFulfilled = (action: FulfilledAction): boolean => action.type.endsWith('/fulfilled');
const isRejected = (action: RejectedAction): boolean => action.type.endsWith('/rejected');

export const isActionPending = (prefix: string) => (action: PendingAction) => {
    return hasPrefix(action, prefix) && isPending(action);
};

export const isActionRejected = (prefix: string) => (action: RejectedAction) => {
    return hasPrefix(action, prefix) && isRejected(action);
};

export const isActionFulfilled = (prefix: string) => (action: FulfilledAction) => {
    return hasPrefix(action, prefix) && isFulfilled(action);
};
