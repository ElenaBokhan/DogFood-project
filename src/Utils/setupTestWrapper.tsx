import {router} from 'App';
import {render} from '@testing-library/react';
import {Provider} from 'react-redux';
import {createMemoryRouter, RouterProvider} from 'react-router-dom';
import {ToastContainer} from 'react-toastify';
import store from 'Store/configureStore';

interface WrapperForTestsParams {
    initialRouterEntries: string[];
}

export const setupTestsWrapper = ({initialRouterEntries}: WrapperForTestsParams) => {
    const memoryRouter = createMemoryRouter(router, {
        initialEntries: initialRouterEntries,
    });
    render(
        <Provider store={store}>
            <RouterProvider router={memoryRouter} />
            <ToastContainer autoClose={2000} position="top-right" theme="light" closeOnClick />
        </Provider>
    );

    return memoryRouter;
};
