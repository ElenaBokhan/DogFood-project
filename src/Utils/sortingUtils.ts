export const calculateOldPrice = (price: number, discount: number) => {
    return Math.floor(price - (price * discount) / 100);
};
