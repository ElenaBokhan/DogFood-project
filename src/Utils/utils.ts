import {PER_PAGE} from 'Const';
import {ESortFilter} from 'Pages/Catalog/Catalog';

export const calculateOldPrice = (price: number, discount: number) => {
    return Math.floor(price - (price * discount) / 100);
};

export const isFavourite = (likes: string[], id: string) => {
    return likes.includes(id);
};

export const getDefaultFilter = (): IClientFilter => {
    return {search: '', page: 1, perPage: PER_PAGE};
};

export const dateFormatter = new Intl.DateTimeFormat('ru-RU', {
    year: 'numeric',
    month: 'short',
    day: '2-digit',
    // hour: '2-digit',
    // minute: '2-digit',
});

export const objectHasProperty = <P extends PropertyKey>(obj: unknown, prop: P): obj is object & Record<P, unknown> => {
    return typeof obj === 'object' && !!obj && Object.hasOwn(obj, prop);
};

export const getResponse = (ms: number) => {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            if (Math.random() > 0.5) {
                resolve('success');
            } else {
                reject(new Error('error'));
            }
        }, ms);
    });
};

export const getSortingProducts = (productList: IProduct[], sortFilter: ESortFilter): IProduct[] => {
    const sortProducts: IProduct[] = [...productList];

    switch (sortFilter) {
        case ESortFilter.POPULAR:
            break;
        case ESortFilter.NEW:
            break;
        case ESortFilter.CHEAPER_FIRST:
            return sortProducts.sort((first, second) => first.price - second.price);
        case ESortFilter.EXPENSIVE_FIRST:
            return sortProducts.sort((first, second) => second.price - first.price);
        case ESortFilter.DISCOUNT:
            return sortProducts.sort((first, second) => second.discount - first.discount);
        case ESortFilter.RATING:
            break;
    }

    return sortProducts;
};
