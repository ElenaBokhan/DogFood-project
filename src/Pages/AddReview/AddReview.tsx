import {Button} from 'Components/Common/Button/Button';
import {Text} from 'Components/Common/Text/Text';
import {TitlePage} from 'Components/Common/TitlePage/TitlePage';
import styles from 'Pages/AddReview/AddReview.module.css';
import {useLocation, useNavigate} from 'react-router-dom';
import starFillIcon from 'assets/ic-star-fill.svg';
import starIcon from 'assets/ic-star.svg';
import {useEffect, useState} from 'react';
import {ETestId} from 'Enum';
import {IconButton} from 'Components/Common/IconButton/IconButton';
import {useAddRewiewMutation} from 'Store/Api/productListApi';
import {Gap} from 'Components/Common/Gap/Gap';

export const AddReview = () => {
    const {state} = useLocation();
    const {name: productName, _id} = state;
    const [addReview, {isSuccess}] = useAddRewiewMutation();
    const navigate = useNavigate();

    const [reviewText, setReviewText] = useState<string>('');
    const [currentRating, setCurrentRating] = useState<number>(null);

    useEffect(() => {
        isSuccess && navigate(`/product/${_id}`);
    }, [_id, isSuccess, navigate]);

    const handleReviewChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
        const reviewText = event.currentTarget.value;
        setReviewText(reviewText);
    };

    const handleSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
        if (currentRating || reviewText) {
            addReview({productId: _id, review: {rating: currentRating, text: reviewText}});
        }
        event.preventDefault();
    };

    const handleSetRating = (rate: number) => {
        setCurrentRating(rate);
    };

    const renderStars = () => {
        return [...Array(5)].map((_, index) => (
            <IconButton
                onClick={() => handleSetRating(index + 1)}
                key={index}
                icon={index < currentRating ? starFillIcon : starIcon}
                alt={'ratingProduct'}
            />
        ));
    };

    const renderAddReviewForm = () => {
        return (
            <form onSubmit={handleSubmit}>
                <div className={styles.field}>
                    <Text value={'Общая оценка'} />
                    {renderStars()}
                </div>
                <div className={styles.field}>
                    <Text value={'Комментарий'} />
                    <input
                        className={styles.reviewTextarea}
                        type="textarea"
                        data-testid={ETestId.ADD_REVIEW_TEXT_INPUT}
                        onChange={handleReviewChange}
                        placeholder={'Поделитесь впечатлениями о товаре'}
                        value={reviewText}
                    />
                </div>
                <Button testId={ETestId.ADD_REVIEW_SUBMIT_BUTTON} type="submit" label={'Отправить отзыв'} />
            </form>
        );
    };

    return (
        <div className={styles.addReviewPage}>
            <TitlePage label={`Отзыв о товаре ${productName}`} />
            <Gap size={40} />
            <hr />
            {renderAddReviewForm()}
        </div>
    );
};
