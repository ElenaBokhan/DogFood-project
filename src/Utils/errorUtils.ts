import {objectHasProperty} from 'Utils/utils';

export const getMessageFromError = (error: unknown, defaultErrorMessage: string) => {
    if (
        objectHasProperty(error, 'data') &&
        objectHasProperty(error.data, 'message') &&
        typeof error.data.message === 'string'
    )
        return error.data.message;
    return defaultErrorMessage;
};
