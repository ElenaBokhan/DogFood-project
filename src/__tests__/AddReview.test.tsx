import '@testing-library/jest-dom';
import {setupTestsWrapper} from 'Utils/setupTestWrapper';
import {render, screen, fireEvent} from '@testing-library/react';
import {AddReview} from 'Pages/AddReview/AddReview';
import {Provider} from 'react-redux';
import store from 'Store/configureStore';
import {BrowserRouter, createMemoryRouter} from 'react-router-dom';
import * as ReactRouter from 'react-router';
import {ETestId} from 'Enum';
import {userEvent} from '@testing-library/user-event';

describe('Тесты для компонента AddReview', () => {
    let router: ReturnType<typeof createMemoryRouter> | undefined;
    let submitBtn: HTMLButtonElement | undefined;
    let textInput: HTMLInputElement | undefined;
    const name = 'productName';
    const productId = '123';

    beforeEach(() => {
        jest.spyOn(ReactRouter, 'useLocation').mockImplementation(() => ({
            state: {name, _id: productId},
            pathname: '',
            hash: '',
            key: '',
            search: '',
        }));
        router = setupTestsWrapper({initialRouterEntries: ['/addReview']});
        submitBtn = screen.getByTestId(ETestId.ADD_REVIEW_SUBMIT_BUTTON);
        textInput = screen.getByTestId(ETestId.ADD_REVIEW_TEXT_INPUT);
    });

    describe('Проверяем элементы компонента AddReview', () => {
        test('Проверяем что поле текстового сообщения и кнопка добавления отзыва доступны', async () => {
            expect(submitBtn).toBeEnabled();
            expect(textInput).toBeEnabled();
        });
        test('Проверяем что запрос не выполняется при незаполненных поле текстового сообщения и редиректа не происходит', async () => {
            expect(textInput).toHaveDisplayValue('');

            await userEvent.click(submitBtn);
            expect(router?.state.location.pathname).toEqual('/addReview');
        });

        test('Проверяем что запрос выполняется при заполненом поле текстового сообщения и происходит редирект на страницу товара', async () => {
            const message = 'Новое сообщение';

            await fireEvent.change(textInput, {target: {value: message}});

            expect(textInput).toHaveDisplayValue(message);

            await userEvent.click(submitBtn);
            expect(router?.state.location.pathname).toEqual(`/product/${productId}`);
        });
    });

    describe('Снапшот компонента AddReview', () => {
        test('Snapshot', () => {
            const {asFragment} = render(
                <Provider store={store}>
                    <BrowserRouter>
                        <AddReview />
                    </BrowserRouter>
                </Provider>
            );

            expect(asFragment()).toMatchSnapshot();
        });
    });
});
