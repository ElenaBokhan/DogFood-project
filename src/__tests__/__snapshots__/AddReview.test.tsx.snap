// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тесты для компонента AddReview Снапшот компонента AddReview Snapshot 1`] = `
<DocumentFragment>
  <div
    class="addReviewPage"
  >
    <div>
      <a
        class="breadcrumbs"
        href="/"
      >
        <img
          alt="arrow"
          src="[object Object]"
        />
        <span
          class="grey undefined subscribe2"
        >
          Назад
        </span>
      </a>
      <span
        class="black undefined head1"
      >
        Отзыв о товаре productName
      </span>
    </div>
    <div
      style="height: 40px;"
    />
    <hr />
    <form>
      <div
        class="field"
      >
        <span
          class="black undefined paragraph1"
        >
          Общая оценка
        </span>
        <button
          class="button"
          type="button"
        >
          <img
            alt="ratingProduct"
            src="[object Object]"
          />
        </button>
        <button
          class="button"
          type="button"
        >
          <img
            alt="ratingProduct"
            src="[object Object]"
          />
        </button>
        <button
          class="button"
          type="button"
        >
          <img
            alt="ratingProduct"
            src="[object Object]"
          />
        </button>
        <button
          class="button"
          type="button"
        >
          <img
            alt="ratingProduct"
            src="[object Object]"
          />
        </button>
        <button
          class="button"
          type="button"
        >
          <img
            alt="ratingProduct"
            src="[object Object]"
          />
        </button>
      </div>
      <div
        class="field"
      >
        <span
          class="black undefined paragraph1"
        >
          Комментарий
        </span>
        <input
          class="reviewTextarea"
          data-testid="addReviewTextInput"
          placeholder="Поделитесь впечатлениями о товаре"
          type="textarea"
          value=""
        />
      </div>
      <button
        class="button standard undefined"
        data-testid="addReviewSubmitButton"
        data-value="Отправить отзыв"
        type="submit"
      >
        Отправить отзыв
      </button>
    </form>
  </div>
</DocumentFragment>
`;
